ENH: add simplified fast numerical inversion to stats.sampling

chrisb83 opened the pull request:
#### Reference issue
n/a

#### What does this implement/fix?
In `stats.sampling`, various methods are implemented to generate random variates based on a specified distribution (the methods are from the UNURAN C library). The tools are mainly for advanced users and are not very easy to use. In this PR, a config is created for a large number of the SciPy distributions such that `NumericalInversePolynomial` can be directly used w/o having to care about the settings. For example, to sample from the Gamma distribution, one can just write

```
rng = FastGeneratorInversion('gamma', (1.3,))
rng.rvs(size=10)
```

Further supported features:
- a fast inverse CDF `ppf_fast` that is computed when the object is created
- sampling with quasi-random numbers from `stats.qmc` (`qrvs`)
- truncation (all distributions can be truncated/restricted to an interval using the keyword `domain`)
- relying on the usual distribution infrastructure, methods like `pdf`, `cdf` are added to the distribution object (this is important in case truncation is used, the existing infrastructure does not support the evaluation of the pdf, cdf)


#### Additional information

- Extensive testing has been performed to ensure that the ppf used for sampling is accurate (more details to be added). To this end, a recommended range of shape parameters has been specified in the config that ensures that the numerical precision of the inversion is controlled.
- For a lot of distributions, the performance of rvs is better than that of `rv_continuous` if large samples are required (for small samples, the cost of the setup step to derive `ppf_fast` is too high compared to the marginal sampling times)
- The functionality is not supposed to replace the existing `rv_continuous` infrastructure: the main idea is to make the UNURAN functionality more easily available and to allow truncations/qrvs for a large number of SciPy distributions.
- Note that the functionality only works with a single set of shape parameters, e.g., for the Gamma distribution, one has to fix a single value of the shape parameter, an array of different shape parameters cannot be used (since for each shape parameter, `ppf_fast` has to be created separately)

Below are some examples (time in ms, generation of 10**7 rvs, statistics are computed over 10 trials):
For the new method, one has to take into account the setup time + the marginal sampling time

----- beta with parameters (2.5, 3.9) -----
New method setup speed (min., mean, std): 31.3, 48.2, 17.5
New method sampling speed (min., mean, std): 469.2, 540.2, 52.1
SciPy method sampling speed (min., mean, std): 1305.7, 1467.1, 137.6

----- crystalball with parameters (0.7, 1.2) -----
New method setup speed (min., mean, std): 66.5, 88.7, 19.8
New method sampling speed (min., mean, std): 399.6, 451.8, 45.2
SciPy method sampling speed (min., mean, std): 2515.3, 2809.9, 236.1

----- gennorm with parameters (2.7,) -----
New method setup speed (min., mean, std): 12.0, 16.8, 2.0
New method sampling speed (min., mean, std): 395.4, 414.5, 17.0
SciPy method sampling speed (min., mean, std): 1798.0, 1933.1, 98.5

<details>
<summary>Detailed performance results</summary>

----- alpha with parameters (3.5,) -----
Intervals (min., mean, std): 160.0, 160.0, 0.0
New method setup speed (min., mean, std): 15.6, 15.7, 0.1
New method sampling speed (min., mean, std): 249.3, 263.8, 8.6
SciPy method sampling speed (min., mean, std): 406.2, 413.3, 7.4

----- anglit with parameters () -----
Intervals (min., mean, std): 137.0, 137.0, 0.0
New method setup speed (min., mean, std): 0.0, 12.5, 6.2
New method sampling speed (min., mean, std): 249.9, 256.3, 7.7
SciPy method sampling speed (min., mean, std): 266.1, 285.4, 20.0

----- argus with parameters (3.5,) -----
Intervals (min., mean, std): 100.0, 100.0, 0.0
New method setup speed (min., mean, std): 0.0, 10.3, 7.0
New method sampling speed (min., mean, std): 249.9, 275.0, 29.8
SciPy method sampling speed (min., mean, std): 562.4, 569.2, 8.2

----- argus with parameters (5.1,) -----
Intervals (min., mean, std): 104.0, 104.0, 0.0
New method setup speed (min., mean, std): 0.0, 11.6, 7.8
New method sampling speed (min., mean, std): 371.8, 378.8, 6.8
SciPy method sampling speed (min., mean, std): 562.4, 579.1, 23.3

----- beta with parameters (0.3, 0.3) -----
Intervals (min., mean, std): 311.0, 311.0, 0.0
New method setup speed (min., mean, std): 194.6, 219.2, 28.2
New method sampling speed (min., mean, std): 325.1, 376.0, 69.4
SciPy method sampling speed (min., mean, std): 921.8, 1121.8, 266.9

----- beta with parameters (1.5, 0.9) -----
Intervals (min., mean, std): 166.0, 166.0, 0.0
New method setup speed (min., mean, std): 70.8, 100.0, 19.3
New method sampling speed (min., mean, std): 421.1, 519.9, 68.0
SciPy method sampling speed (min., mean, std): 1623.1, 1853.5, 132.3

----- beta with parameters (2.5, 3.9) -----
Intervals (min., mean, std): 135.0, 135.0, 0.0
New method setup speed (min., mean, std): 31.3, 48.2, 17.5
New method sampling speed (min., mean, std): 469.2, 540.2, 52.1
SciPy method sampling speed (min., mean, std): 1305.7, 1467.1, 137.6

----- cosine with parameters () -----
Intervals (min., mean, std): 110.0, 110.0, 0.0
New method setup speed (min., mean, std): 0.0, 8.8, 8.0
New method sampling speed (min., mean, std): 298.7, 310.9, 10.0
SciPy method sampling speed (min., mean, std): 1132.3, 1506.9, 375.7

----- betaprime with parameters (2.5, 3.3) -----
Intervals (min., mean, std): 137.0, 137.0, 0.0
New method setup speed (min., mean, std): 31.2, 34.9, 4.6
New method sampling speed (min., mean, std): 285.2, 303.0, 11.1
SciPy method sampling speed (min., mean, std): 1016.7, 1032.3, 9.8

----- betaprime with parameters (0.3, 0.3) -----
Intervals (min., mean, std): 311.0, 311.0, 0.0
New method setup speed (min., mean, std): 184.2, 243.1, 51.6
New method sampling speed (min., mean, std): 283.1, 334.5, 51.2
SciPy method sampling speed (min., mean, std): 1648.3, 1671.9, 22.7

----- betaprime with parameters (1.5, 0.9) -----
Intervals (min., mean, std): 166.0, 166.0, 0.0
New method setup speed (min., mean, std): 48.9, 58.4, 8.0
New method sampling speed (min., mean, std): 283.0, 306.2, 10.7
SciPy method sampling speed (min., mean, std): 1364.3, 1398.6, 16.8

----- bradford with parameters (1.2,) -----
Intervals (min., mean, std): 9.0, 9.0, 0.0
New method setup speed (min., mean, std): 0.0, 7.1, 8.2
New method sampling speed (min., mean, std): 314.1, 317.5, 4.8
SciPy method sampling speed (min., mean, std): 399.6, 417.9, 9.0

----- burr with parameters (1.3, 2.4) -----
Intervals (min., mean, std): 158.0, 158.0, 0.0
New method setup speed (min., mean, std): 33.1, 42.5, 7.4
New method sampling speed (min., mean, std): 290.7, 303.7, 8.5
SciPy method sampling speed (min., mean, std): 731.6, 769.4, 23.0

----- burr12 with parameters (0.7, 1.2) -----
Intervals (min., mean, std): 174.0, 174.0, 0.0
New method setup speed (min., mean, std): 31.7, 44.8, 7.7
New method sampling speed (min., mean, std): 282.9, 301.4, 11.7
SciPy method sampling speed (min., mean, std): 951.2, 981.6, 16.6

----- cauchy with parameters () -----
Intervals (min., mean, std): 203.0, 203.0, 0.0
New method setup speed (min., mean, std): 15.6, 19.8, 6.5
New method sampling speed (min., mean, std): 299.7, 311.7, 10.4
SciPy method sampling speed (min., mean, std): 300.7, 312.8, 6.1

----- chi2 with parameters (3.5,) -----
Intervals (min., mean, std): 156.0, 156.0, 0.0
New method setup speed (min., mean, std): 15.7, 30.3, 5.2
New method sampling speed (min., mean, std): 284.1, 299.4, 11.8
SciPy method sampling speed (min., mean, std): 467.2, 487.8, 10.3

----- chi with parameters (4.5,) -----
Intervals (min., mean, std): 121.0, 121.0, 0.0
New method setup speed (min., mean, std): 16.6, 22.7, 6.6
New method sampling speed (min., mean, std): 428.8, 530.9, 152.9
SciPy method sampling speed (min., mean, std): 565.1, 612.8, 35.7

----- crystalball with parameters (0.7, 1.2) -----
Intervals (min., mean, std): 272.0, 272.0, 0.0
New method setup speed (min., mean, std): 66.5, 88.7, 19.8
New method sampling speed (min., mean, std): 399.6, 451.8, 45.2
SciPy method sampling speed (min., mean, std): 2515.3, 2809.9, 236.1

----- expon with parameters () -----
Intervals (min., mean, std): 76.0, 76.0, 0.0
New method setup speed (min., mean, std): 0.0, 10.4, 7.2
New method sampling speed (min., mean, std): 398.8, 444.1, 34.8
SciPy method sampling speed (min., mean, std): 282.0, 326.3, 48.4

----- gamma with parameters (1.5,) -----
Intervals (min., mean, std): 117.0, 117.0, 0.0
New method setup speed (min., mean, std): 14.7, 26.9, 7.5
New method sampling speed (min., mean, std): 383.1, 476.4, 56.3
SciPy method sampling speed (min., mean, std): 709.8, 869.2, 115.6

----- geninvgauss with parameters (3.3, 1.1) -----
Intervals (min., mean, std): 123.0, 123.0, 0.0
New method setup speed (min., mean, std): 16.7, 28.0, 7.3
New method sampling speed (min., mean, std): 399.6, 436.5, 31.9
SciPy method sampling speed (min., mean, std): 1981.2, 2111.2, 50.0

----- gennorm with parameters (2.7,) -----
Intervals (min., mean, std): 124.0, 124.0, 0.0
New method setup speed (min., mean, std): 12.0, 16.8, 2.0
New method sampling speed (min., mean, std): 395.4, 414.5, 17.0
SciPy method sampling speed (min., mean, std): 1798.0, 1933.1, 98.5

----- gumbel_l with parameters () -----
Intervals (min., mean, std): 123.0, 123.0, 0.0
New method setup speed (min., mean, std): 0.0, 14.9, 8.2
New method sampling speed (min., mean, std): 398.5, 461.4, 49.6
SciPy method sampling speed (min., mean, std): 613.9, 689.5, 51.8

----- gumbel_r with parameters () -----
Intervals (min., mean, std): 127.0, 127.0, 0.0
New method setup speed (min., mean, std): 3.2, 12.7, 8.9
New method sampling speed (min., mean, std): 401.3, 457.1, 49.0
SciPy method sampling speed (min., mean, std): 416.3, 476.3, 39.8

----- hypsecant with parameters () -----
Intervals (min., mean, std): 132.0, 132.0, 0.0
New method setup speed (min., mean, std): 12.0, 21.6, 15.4
New method sampling speed (min., mean, std): 401.6, 492.5, 71.1
SciPy method sampling speed (min., mean, std): 481.8, 536.1, 43.3

----- invgauss with parameters (3.1,) -----
Intervals (min., mean, std): 137.0, 137.0, 0.0
New method setup speed (min., mean, std): 15.7, 30.9, 10.6
New method sampling speed (min., mean, std): 300.6, 330.2, 18.1
SciPy method sampling speed (min., mean, std): 517.2, 662.7, 137.6

----- invweibull with parameters (1.5,) -----
Intervals (min., mean, std): 158.0, 158.0, 0.0
New method setup speed (min., mean, std): 16.0, 28.0, 8.2
New method sampling speed (min., mean, std): 413.6, 461.5, 44.4
SciPy method sampling speed (min., mean, std): 767.0, 872.0, 89.7

----- laplace with parameters () -----
Intervals (min., mean, std): 141.0, 141.0, 0.0
New method setup speed (min., mean, std): 1.0, 14.8, 6.7
New method sampling speed (min., mean, std): 346.5, 489.8, 70.5
SciPy method sampling speed (min., mean, std): 314.3, 408.4, 37.7

----- logistic with parameters () -----
Intervals (min., mean, std): 129.0, 129.0, 0.0
New method setup speed (min., mean, std): 14.0, 18.2, 5.1
New method sampling speed (min., mean, std): 392.3, 428.0, 25.2
SciPy method sampling speed (min., mean, std): 333.0, 363.0, 19.0

----- maxwell with parameters () -----
Intervals (min., mean, std): 120.0, 120.0, 0.0
New method setup speed (min., mean, std): 0.0, 10.9, 7.9
New method sampling speed (min., mean, std): 400.7, 452.3, 41.3
SciPy method sampling speed (min., mean, std): 781.6, 844.1, 52.5

----- moyal with parameters () -----
Intervals (min., mean, std): 127.0, 127.0, 0.0
New method setup speed (min., mean, std): 1.0, 13.6, 5.6
New method sampling speed (min., mean, std): 299.6, 382.7, 105.7
SciPy method sampling speed (min., mean, std): 1382.9, 1587.6, 157.2

----- norm with parameters () -----
Intervals (min., mean, std): 124.0, 124.0, 0.0
New method setup speed (min., mean, std): 3.7, 12.2, 5.4
New method sampling speed (min., mean, std): 345.8, 399.6, 41.5
SciPy method sampling speed (min., mean, std): 283.0, 301.5, 21.5

----- pareto with parameters (1.3,) -----
Intervals (min., mean, std): 111.0, 111.0, 0.0
New method setup speed (min., mean, std): 13.0, 16.3, 3.4
New method sampling speed (min., mean, std): 455.9, 491.4, 32.1
SciPy method sampling speed (min., mean, std): 716.3, 760.3, 36.5

----- powerlaw with parameters (7.6,) -----
Intervals (min., mean, std): 66.0, 66.0, 0.0
New method setup speed (min., mean, std): 5.8, 7.2, 1.3
New method sampling speed (min., mean, std): 412.5, 457.2, 46.6
SciPy method sampling speed (min., mean, std): 659.0, 731.7, 62.0

----- rayleigh with parameters () -----
Intervals (min., mean, std): 142.0, 142.0, 0.0
New method setup speed (min., mean, std): 9.9, 10.7, 0.8
New method sampling speed (min., mean, std): 281.4, 303.9, 12.6
SciPy method sampling speed (min., mean, std): 340.0, 375.7, 42.4

----- semicircular with parameters () -----
Intervals (min., mean, std): 92.0, 92.0, 0.0
New method setup speed (min., mean, std): 9.0, 11.4, 3.6
New method sampling speed (min., mean, std): 293.4, 308.3, 12.3
SciPy method sampling speed (min., mean, std): 385.2, 399.8, 12.9

----- t with parameters (5.7,) -----
Intervals (min., mean, std): 141.0, 141.0, 0.0
New method setup speed (min., mean, std): 10.0, 11.5, 1.0
New method sampling speed (min., mean, std): 287.8, 298.9, 10.3
SciPy method sampling speed (min., mean, std): 780.5, 785.9, 4.0

----- wald with parameters () -----
Intervals (min., mean, std): 133.0, 133.0, 0.0
New method setup speed (min., mean, std): 13.0, 14.7, 1.5
New method sampling speed (min., mean, std): 288.2, 303.4, 11.4
SciPy method sampling speed (min., mean, std): 512.6, 520.2, 5.5

----- weibull_max with parameters (2.4,) -----
Intervals (min., mean, std): 112.0, 112.0, 0.0
New method setup speed (min., mean, std): 11.9, 13.8, 2.2
New method sampling speed (min., mean, std): 289.2, 306.5, 13.9
SciPy method sampling speed (min., mean, std): 521.9, 537.6, 9.3

----- weibull_min with parameters (1.2,) -----
Intervals (min., mean, std): 111.0, 111.0, 0.0
New method setup speed (min., mean, std): 15.0, 16.6, 1.2
New method sampling speed (min., mean, std): 286.2, 308.4, 16.6
SciPy method sampling speed (min., mean, std): 675.7, 692.5, 8.2

</details>

It is important to remark that `ppf_fast` is typically quite a bit faster than `ppf`, so for `qrvs`, the speedup should be substantial compared to using `ppf`. Example for the Student t distribution:

```
import numpy as np
from scipy import stats
from scipy import special as sc


urng = np.random.default_rng()
df = 1.5
u = urng.uniform(size=1_000_000)

# use ppf:
%timeit stats.t.ppf(u, df)
# 2.54 s ± 22.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

# use ppf_fast (here we define the generator with NumericalInversePolynomial directly, its ppf is the same as ppf_fast)
gen = stats.sampling.NumericalInversePolynomial(stats.t(df), center=0)

%timeit gen.ppf(u)
# 28.1 ms ± 2.41 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

chrisb83 commented on scipy/stats/sampling.py:
```diff
@@ -49,3 +70,1262 @@
+
+    Parameters
+    ----------
+    distname : str
```
I need to add the list of distributions that are supported (the keys of the config dictionary). TBD how this can be achieved (f-strings cannot be used for the doc-string)

mdhaber replied:
`_lib/_docscrape.py` can help to modify docstrings programmatically. Some example uses are in `stats/_axis_nan_policy.py`.

WarrenWeckesser replied:
But please be careful with docstring hacking--see https://github.com/scipy/scipy/issues/17189.

tupui commented on scipy/stats/sampling.py:
```diff
@@ -49,3 +70,1262 @@
+        The random state used in relevant methods like `rvs` (unless
+        another `random_state` is passed as an argument to these methods).
+
+    Methods
+    -------
+    cdf
+    evaluate_error
+    isf
+    logpdf
+    pdf
+    ppf
+    ppf_fast
+    qrvs
+    rvs
+    set_loc_scale
+    support
+    sf
```
I am concerned about the API. What is the scope of this class? As it is, as a users, this would confuse me with distributions, frozen dists and this.

I really think we should focus on the sampling part, as the description suggests. i.e. here only propose `rvs, qrvs, evaluate_error`. If needed internally, others things should be private. 

For `rvs, qrvs` ideally I would prefer to only have a single method and set this during the init. But since we already have this API elsewhere I suppose it's ok. Another though is that I am not sure if its useful to keep a non QMC option.

chrisb83 replied:
I think pdf, cdf, ppf are important to have, otherwise you always need to create another object from rv_continuous to have access to these methods. For example, if you want to create a histogram, you need to do something like

```
x = np.linspace(0, 1, 100)
rng1 = FastGeneratorInversion('beta', (1.5, 2.5))
rng2 = stats.beta(1.5, 2.5)
r = rng1.rvs(size=1000)
y = rng2.pdf(x)
...
```

and if you use truncation, we do not support that in rv_continuous yet, so the user would manually have to compute the pdf/cdf

chrisb83 replied:
I would agree to removing the following methods: `isf`, `logpdf`, `sf`

`sf` and `isf` might be required in some situations but probably not very often, `logpdf` might be most relevant for MLE and `fit` is not supported anyways. this would already simplify the API quite a bit while keeping the key functions that are required in the context of sampling

chrisb83 replied:
I also discussed this topic with Matt. There are mainly four options:

1. implement all methods that we have in a frozen object for consistency, e.g., `moment`, `logpdf`, etc including support for truncation
2. focus on sampling including basic support for `pdf` and `cdf` (to create histograms, run a goodness-of-fit test)
3. focus on the sampling, i.e., keep only `rvs`, `qrvs`, `ppf_fast`, `evaluate_error`. `pdf`/`cdf` could be added later if  requested by users.
4. integrate the approach directly into `sampling.NumericalInversePolynomial`, i.e., if a frozen SciPy dist is passed, the settings proposed in this PR are used to create the generator

From my point of view, 1. is too much extra work and there is too much overlap with the strategic API that it intended to support truncation.
Since there is not enough support for 2., I would propose option 3.

After thinking a bit more about option 4, I don't think it is a good idea for various reasons:
- both the documentation and implementation will become confusing with many case distinctions that describe the different behaviors. this PR was intended to make sampling methods more accessible to users who are not experts in random variate generation, while keeping full flexibility in the original `sampling.NumericalInversePolynomial`. Also from this perspective, it is easier to keep them separated.
- I would prefer to minimize changes to the UNU.RAN integration. for example, `sampling.NumericalInversePolynomial` does not even have a separate `rvs` method, all generators use a generic rvs methdod in `stats/_unuran/unuran_wrapper.pyx` which is nice from a consistency point of view.

tupui commented on scipy/stats/sampling.py:
```diff
@@ -49,3 +70,1262 @@
+
+        return CustomDistPINV(cfg["pdf"], args)
+
+    def rvs(self, size=None, random_state=None):
```
Please remove `random_state`. We have a class, hence a state.

chrisb83 replied:
I agree, if needed, the attribute random_state can be changed by the user

tupui commented on scipy/stats/sampling.py:
```diff
@@ -49,3 +70,1262 @@
+        # update the adjusted domain that depends on loc and scale
+        self._set_domain_adj()
+
+    def evaluate_error(self, size=100000, random_state=None, x_error=False):
```
Ideally it should be 99999 as with `bootstrap`

tupui commented on scipy/stats/sampling.py:
```diff
@@ -49,3 +70,1262 @@
+                qrvs = qrvs.reshape(tuple_size + (d,))
+        return loc + scale * qrvs
+
+    def set_loc_scale(self, loc=0, scale=1):
```
I would avoid using `set`. It does not sound very Pythonic. Why not storing both separately as attributes? (then use properties)

chrisb83 replied:
if you change loc/scale, you also need to update the frozen dist / domain in the background, that is why I did not define them as attributes

chrisb83 replied:
done, I will mark this as resolved

tupui commented on scipy/stats/sampling.py:
```diff
@@ -39,7 +39,28 @@
+   FastGeneratorInversion
+
 """
+import math
```
What is the reason for not using NumPy?

chrisb83 replied:
Speed, I added a comment in the description of the dictionary:

> pdf: the pdf of the distribution (callable). The signature of the pdf is float -> float (i.e., the function does not have to be vectorized). If possible, functions like log or exp from the module math should be preferred over functions from numpy since the PINV setup will be faster in that case.

I think the difference is quite substantial, ~3x/4x for the setup, I tested it a while ago, I will add details later

tupui commented on scipy/stats/sampling.py:
```diff
@@ -49,3 +70,1262 @@
+
+# define pdfs and other helper functions to create the generators
+
+def argus_pdf(x, chi):
```
These all look quite simple. Is it an option to write them with Pythran or Cython?

chrisb83 replied:
that would indeed be a big advantage, it would make the setup much faster. but it requires a change of the interface of NumericalInversePolynomial: it needs to support LowLevelCallable (I think Tirth left a comment about this in the code). or one has to work directly with the UNURAN C code. this could be a nice follow-up project

tupui reviewed (commented):
Thank you for the work @chrisb83 🚀 I wrote some points to get us started.

chrisb83 commented on scipy/stats/sampling.py:
```diff
@@ -39,7 +39,28 @@
+
 """
+import math
+import warnings
```
I have to move the code to a separate file `_sampling.py` to fix the test failures. I think I tried this but `from ._sampling import *` did not work in `sampling.py`.

chrisb83 replied:
fixed by updating meson.build (thanks @tupui)

tupui commented on scipy/stats/_sampling.py:
```diff
@@ -1051,23 +1043,15 @@ def evaluate_error(self, size=100000, random_state=None, x_error=False):
-            urng.set_state(state)
-        elif isinstance(urng, np.random.Generator):
-            urng.bit_generator.state = state
+        # urng will be used to draw the samples for testing the error
+        # it must not interfere with self.random_state. therefore, do not
+        # call self.rvs, but draw uniform random numbers and apply
+        # self.ppf_fast (note: like in rvs, consider self._mirror_uniform)
```
To me this sounds like an implementation detail. Why not just creating another rng, based on the provided rng, when the class is instantiated?

tupui replied:
What I mean is that IMO we should not ask the user to handle 2 RNG but do this ourselves internally. We can spawn another RNG based on the first one. i.e. we would have another private attribute for this RNG.

As for given the user the flexibility to update the value. I'm still not convinced about the need to be able to update the "main" RNG. Having the possibility to update this one is even more nebulous to me.

chrisb83 replied:
for the evaluation of the error, it is important to leave the control to the user to make the calculation reproducible. this should not be a state in my view, here is really seems natural to say "estimate the error with seed XYZ"
we can take this up in the context of the other RNG topic :)

tupui commented on scipy/stats/_sampling.py:
```diff
@@ -735,6 +735,14 @@ def __init__(
         else:
             raise ValueError(f"Unknown distname: {distname}")
 
+    @property
+    def random_state(self):
+        return self._random_state
+
+    @random_state.setter
+    def random_state(self, random_state):
+        self._random_state = check_random_state(random_state)
```
Is there a use case we really want to support that motivates having this as public? I would like to keep the public API as lean as possible.

chrisb83 replied:
After removing `random_state` from `rvs`, I thought it is important to allow the user to change the seed easily. For example, it could be that you create the generator, then check that the rvs are indeed ok (maybe a histogram or GoF test) before you start your actual simulation with a fixed seed.
before the change, you could have done this without `self.random_state` by calling `rvs(size=..., random_state=123)` for the actual simulation

tupui replied:
What you describe is more like a `reset` functionality no? This is what we do in `QMCEngine.reset` for instance.

If the workflow requires to first get some samples and then run "for real", then yes this would be an issue with QMC and IMO `reset` or something else would be clearer for the user than saying update the RNG.

rkern replied:
Unfortunately, I think the `QMCEngine.reset()` semantics do not work well for dealing with the PRNG state ([as we've noted elsewhere](https://github.com/scipy/scipy/pull/16923#issuecomment-1234551154)). They work okay when the user passes in a static seed, but not when a user passes in a live `Generator` instance. You end up with situations where the user can easily create false correlations by unintentionally sampling the same parts of the PRNG state that they did not intend to (i.e. separate from the _intended_ resampling of the `reset()`). Or worse, not reset correctly at all.

```python
|85> rng = np.random.default_rng(0xa29cabb11cfdf44ff6cac8bec254c2a0)

|86> qrng = qmc.LatinHypercube(2, seed=rng)

|87> rng.random(4)  # Use the Generator a little after we've initialized the QMCEngine
array([0.24573226, 0.85685711, 0.02918954, 0.96401299])

|89> qrng.random(8)
array([[0.31600266, 0.17649214],
       [0.65166008, 0.89000517],
       [0.15227553, 0.38101248],
       [0.92677475, 0.10827375],
       [0.81506436, 0.79478042],
       [0.61868508, 0.73785412],
       [0.05790777, 0.29385173],
       [0.38745175, 0.57504755]])

|90> qrng.reset()
<scipy.stats._qmc.LatinHypercube at 0x7f856c4cad00>

|91> qrng.random(8)
array([[0.21928347, 0.26789286],
       [0.49635131, 0.87949838],
       [0.81600266, 0.17649214],
       [0.65166008, 0.64000517],
       [0.27727553, 0.50601248],
       [0.05177475, 0.10827375],
       [0.94006436, 0.41978042],
       [0.61868508, 0.86285412]])
```

I would probably go back to passing it in as an argument to the method. It's fine! It's a pattern that works well. It's okay to have pretty dumb classes that bundle static _data_ and behavior and not _state_ and behavior.

But if you really want this class to hold onto and manage its `Generator` state, then you do have to actually expose it like this so that the user can work with it. Making a class responsible for managing state (not just data) does have consequences for complicating its API.

tupui replied:
> Unfortunately, I think the QMCEngine.reset() semantics do not work well for dealing with the PRNG state (https://github.com/scipy/scipy/pull/16923#issuecomment-1234551154). They work okay when the user passes in a static seed, but not when a user passes in a live Generator instance.

I will just fix that. Somehow this went through the cracks.

Let's get some other opinions for this side of the API then. (With the assumption that the issue you pointed out would be fixed.)

chrisb83 replied:
One additional point: even if this is fixed, a reset method would not be very flexible, e.g., you cannot switch from one generator to another or change the seed. so the API would not be simpler (one method vs one attribute) but less powerful.

rkern replied:
No, I think it's more of a fundamental problem, not an implementation issue. You can't just reset `Generator` to an earlier state safely except in very strictly controlled circumstances (that would not be appropriate to impose on a general purpose API like this one).

I'd prefer to just keep the API minimal, and pass in `random_state=` arguments.

tupui replied:
Ok I had another quick look. What is exactly the issue with `reset`? Because your example here does not show an issue to me. What did you expect? (Also running twice a LHS is a bad idea in general as people could be tempted to concatenate that and the result is not anymore a LHS.)

For the Halton case, I remember why there is a copy of the RNG. This is because we want the same permutation for the Van der Corput sequence (the `rng.shuffle(perm)` part.) This is a must and failing to do that result in a loss of convergence speed (cf. some articles from Art Owen and I also could verify it. I found some emails where we discussed all that when I implemented this.)

rkern replied:
I expected it to reset and reproduce the sequence of samples like it does elsewhere:

```python
|92> rng = np.random.default_rng(0xa29cabb11cfdf44ff6cac8bec254c2a0)

|93> qrng = qmc.LatinHypercube(2, seed=rng)

|94> qrng.random(8)
array([[0.21928347, 0.26789286],
       [0.49635131, 0.87949838],
       [0.81600266, 0.17649214],
       [0.65166008, 0.64000517],
       [0.27727553, 0.50601248],
       [0.05177475, 0.10827375],
       [0.94006436, 0.41978042],
       [0.61868508, 0.86285412]])

|95> qrng.reset()
<scipy.stats._qmc.LatinHypercube at 0x7f856c1a7b50>

|96> qrng.random(8)
array([[0.21928347, 0.26789286],
       [0.49635131, 0.87949838],
       [0.81600266, 0.17649214],
       [0.65166008, 0.64000517],
       [0.27727553, 0.50601248],
       [0.05177475, 0.10827375],
       [0.94006436, 0.41978042],
       [0.61868508, 0.86285412]])
```

I will leave the use case of having a `reset()` method at to you; I assume there is a use case for `reset()` since you implemented it and are advocating for it again here. I'm just pointing out that it fails to do what it claims in this case.

We can talk about the details of the `Halton` case elsewhere (I don't think it's true that the permutations are the same because they are of sequences of different sizes), but even if you do need N identical parallel streams to pass to N workers, you can't just copy the original PRNG state N times to get that because the original stream won't advance. That means that you are going to be resampling the same parts of the PRNG stream even for places where you didn't want that (i.e. between a worker and the original caller). And that's the general problem with any kind of resetting semantics on top of a PRNG. You make it really easy for the user to shoot themselves in the foot. It's actually safer for them to handle the `Generator` themselves rather than try to hide it from them.

If you want a simple, minimal API, let users pass in the `random_state=` as an argument when they call `rvs()`. Any kind of PRNG state management is a complication that we don't need here.

tupui replied:
You will have to be more specific. So far you showed examples of things I am expecting to see. In the last example the sample is the same after the reset. And before that yes there is a difference because you advanced the sample. And one LHS cannot be compared to another smaller or bigger one and won't have the same elements in them. So I don't see what you want to point out with these.

For Halton, the sequence are of the same size.

But yes, let's continue this part of the conversation on an issue if you think there is an issue.

rkern replied:
Yes, we can talk about the details of `QMCEngine.reset()` in other issues. I do strongly contend that they are blockers for adding similar `reset()` methods here. In any case, it doesn't seem like anyone else has any desire for that style of API here (whether or not it can be implemented correctly and safely). No one else wants this class to take on the responsibility of managing that state.

If you object to exposing the stored `.random_state` attribute here in favor of a truly minimal API, that's fine. I recommend keeping it in the method arguments like most other things that randomly sample.

tupui replied:
> it doesn't seem like anyone else has any desire for that style of API here (whether or not it can be implemented correctly and safely). No one else wants this class to take on the responsibility of managing that state.

Let's not jump to conclusions that quickly. This thread is barely a day old. In general this PR will need to see more eyes.

I am still -1 for having `random_state` and `reset` was just a proposal (and there might be other options) if you really want to give some sort of control to users. Which I am still not convinced is needed.

chrisb83 replied:
I will write an email to the mailing list this weekend to ask for further opinions.
I see mainly two options: 

- have a public attribute 
- add random_state as a keyword to rvs, 

I just think it is important to have an API that allows users to change random_state. while not every user will need this, I don't think it complicates the API a lot, one can simply ignore this optional parameter.

chrisb83 replied:
and good to see that the discussion helped to find a bug in SciPy :)

chrisb83 commented:
Two build failures need to be fixed:

- for the Meson build, I have to add the new py-file in stats/tests to meson.build
- the other test failure is about suppressing warnings:

```
E   AssertionError: warning ignore filter should not be used, instead, use
E   numpy.testing.suppress_warnings (in tests only);
E   found in:
E       stats/_sampling.py:738
```

The following information is in the test:

> We try not to add filters in the code base, because those filters aren't thread-safe. We aim to only filter in tests with np.testing.suppress_warnings. However, in some cases it may prove necessary to filter out warnings, because we can't (easily) fix the root cause for them and we don't want users to see some warnings when they use SciPy correctly. So we list exceptions here.  Add new entries only if there's a good reason.

In _sampling.py, I am trying to suppress warnings that can occur when calling `NumericalInversePolynomial` even if the shape parameters are taken from the tested range. So in this sense, it is valid input that would trigger a warning. I can also remove the suppress warnings code and document this behaviour. Thoughts?

rgommers commented on scipy/stats/_sampling.py:
```diff
@@ -0,0 +1,1234 @@
+    gumbel_l, gumbel_r, hypsecant, invgamma, invgauss,
+    invweibull, laplace, logistic, maxwell, moyal,
+    norm, pareto, powerlaw, t, rayleigh, semicircular,
+    wald, weibull_max, weibull_min.
```
It would be good to typeset these as either math (double backticks) or links (single backticks). That's anyway helpful, and also the underscores in the weibull distributions are special reST symbols.

chrisb83 replied:
Ok, I will use double backticks (to make links work, I think I would need to write `scipy.stats.distname` since this file is in `scipy.stats.sampling`, which would be harder to read)

rgommers commented on scipy/stats/_sampling.py:
```diff
@@ -0,0 +1,1234 @@
+
+        if not isinstance(distname, str):
+            raise TypeError("distname must be a string.")
+        if distname not in set(_distn_names):
```
Why not validate this immediately against `PINV_CONFIG`? That's now done further down, which is (a) double validation, and (b) may not be fully robust for the intermediate code.

chrisb83 commented:
@tupui while we are waiting for further feedback, it would be great if you could review the documentation or the test cases. Thanks

tupui commented on scipy/stats/_sampling.py:
```diff
@@ -0,0 +1,1213 @@
+    scale : float, optional
+        The scale parameter. Note that an array of floatsis not allowed.
+        The default is 1.
+    ignore_shape_range : boolean, optional.
+        If False, shape parameters that are outside of the valid range
+        of values to ensure that the numerical accuracy (see Notes) is
+        high, raise a ValueError. If True, any shape parameters that are valid
+        for the distribution are accepted. This can be useful for testing.
+        The default is False.
```
I personally don't like such things and would leave out "testing helpers".

tupui replied:
Can you comment on that? Sorry if we went over it, I don't recall.

chrisb83 replied:
we have tested the numerical error and determined a region in which the error is below the target 1e-10 for each distribution in the config. if a parameter is outside this range, a value error is raised unless you set this variable to True. 

alternatively one could raise a warning and get rid of the ignore_shape_range parameter.

tupui commented on scipy/stats/_sampling.py:
```diff
@@ -0,0 +1,1213 @@
+
+    def evaluate_error(self, size=100000, random_state=None, x_error=False):
+        """
+        Evaluate the numerical accuracy of the inversion (u- and x-error).
```
The description in the note section is missing a layman explanation of what these quantity represent.

chrisb83 replied:
agree, this is too technical

tupui commented on scipy/stats/_sampling.py:
```diff
@@ -0,0 +1,1213 @@
+        Notes
+        -----
+
+        Note that the support of the distribution depends on `loc`,
+        `scale` and `domain`.
```
A figure explaining all that in the docstring's class (or elsewhere) would help.

tupui commented on scipy/stats/_sampling.py:
```diff
@@ -0,0 +1,1213 @@
+        (2.5, 3.5)
+
+        """
+        return self._domain_adj
```
Looks like it could be a property too. It would make it easier to follow the code to have a single thing to keep track of.

tupui commented on scipy/stats/tests/test_fast_gen_inversion.py:
```diff
@@ -0,0 +1,450 @@
+        FastGeneratorInversion(dist)
+
+
+def test_random_state():
```
Looks like you could parametrize these.

tupui commented on scipy/stats/tests/test_fast_gen_inversion.py:
```diff
@@ -0,0 +1,450 @@
+
+        size = (3, 4)
+        d = 5
+        qrng = stats.qmc.Halton(d, seed=0)
+        qrng2 = stats.qmc.Halton(d, seed=0)
```
Or just use `scramble=False` 😉

tupui reviewed (commented):
I did another pass. I think there are still a lot to discuss in terms of API before really diving much into the test part. We need to make some decisions here if we want to move this forward.

chrisb83 commented on scipy/stats/_sampling.py:
```diff
@@ -0,0 +1,1213 @@
+
+    >>> trunc_norm = FastGeneratorInversion('norm', domain=(3, 4))
+    >>> r = trunc_norm.rvs(size=1000)
+    >>> 3 < r.min() < r.max() < 4
+    True
+
+    Check the mean:
+
+    >>> r.mean()
+    3.250433367078603  # may vary
+
+    >>> stats.norm.expect(lb=3, ub=4, conditional=True)
+    3.260454285589997
```
I can do that. Note that the PDF of the truncated distribution would be nice to have if I show a histogram ;)

chrisb83 commented on scipy/stats/_sampling.py:
```diff
@@ -0,0 +1,1213 @@
+    >>> trunc_norm.ppf_fast(0.5)
+    3.198474131740042
+
+    In this particular, case, `scipy.stats.truncnorm` could also be used to
+    generate truncated normal random variates. We can verify that
+    the CDFs coincide up to a small numerical error:
+
+    >>> trunc_norm.cdf(3.5) - stats.truncnorm(3, 4).cdf(3.5)
+    -1.0325074129013956e-14
```
agree, i will remove this

chrisb83 commented on scipy/stats/_sampling.py:
```diff
@@ -0,0 +1,1213 @@
+    >>> r.mean()
+    1.4996639255942914  # may vary
+
+    Compare the PPF against approximation ppf_fast.
+
+    >>> q = [0.001, 0.2, 0.5, 0.8, 0.999]
+    >>> np.max(np.abs(gamma_dist.ppf(q) - gamma_dist.ppf_fast(q)))
+    4.313394796895409e-08
```
I can emphasize better that ppf is the "usual" ppf that aims to be accurate whereas ppf_fast is the ppf designed for sampling where speed is the main consideration. ppf_fast is still very accurate and comparing them is important for the error estimation.

tupui replied:
Having 2 methods would still be redundant and a duplicate. If users want the normal ppf, they can use the original dist.

But I will let other comment on the whole topic :)

chrisb83 commented on scipy/stats/_sampling.py:
```diff
@@ -0,0 +1,1213 @@
+    >>> stats.norm.expect(lb=3, ub=4, conditional=True)
+    3.260454285589997
+
+    Compare `ppf` and `ppf_fast`:
+
+    >>> trunc_norm.ppf(0.5)
+    3.1984741317619285
+    >>> trunc_norm.ppf_fast(0.5)
+    3.198474131740042
```
ok, I will remove it

chrisb83 commented on scipy/stats/_sampling.py:
```diff
@@ -0,0 +1,1213 @@
+        random_state=None,
+    ):
+
+        if not isinstance(distname, str):
+            raise TypeError("distname must be a string.")
+        if distname not in PINV_CONFIG.keys():
+            raise ValueError(f"Distribution '{distname}' is not supported.")
```
Good point, I will join these checks. Do we really want to show the long list of available distributions?

tupui replied:
I personally would. And I don't think we want to have a more complicated logic saying "maybe you meant ..." with an intelligent selection.

chrisb83 replied:
ok, i added PINV_CONFIG.keys()

chrisb83 commented on scipy/stats/_sampling.py:
```diff
@@ -0,0 +1,1213 @@
+        nargs_expected = len(self._frozendist.args)
+        if nargs != nargs_expected:
+            raise ValueError(
+                f"Length of args must be {nargs_expected}, but it is {nargs}"
```
Most distributions just have a single argument, some like beta and betaprime have two. I would say it is acceptable to tell the user how many parameters are expected.

tupui replied:
What I suggest is to tell exactly which args are missing. If you try with 5 and it tells you it needs 2, that's not super helpful if you don't know which ones.

chrisb83 commented on scipy/stats/_sampling.py:
```diff
@@ -0,0 +1,1213 @@
+
+        if domain is None:
+            self._domain = self._frozendist.support()
+            self._p_lower = 0.0
+            self._sf_lower = 0.0
+            self._p_domain = 1.0
+            self._logp_domain = 0.0
```
yes, this just mean that no scaling of the CDF,PDF etc is required as long as there is no truncation. w/o truncation, you can just rely on the regular methods

chrisb83 commented on scipy/stats/_sampling.py:
```diff
@@ -0,0 +1,1213 @@
+            elif self._center > self._domain_pinv[1]:
+                self._center = self._domain_pinv[1]
+
+        self.rng = NumericalInversePolynomial(
```
agree

chrisb83 commented on scipy/stats/_sampling.py:
```diff
@@ -0,0 +1,1213 @@
+                    msg += "with the selected values."
+                    raise ValueError(msg)
+
+        if "center" in cfg.keys():
+            if not np.isscalar(cfg["center"]):
+                self._center = cfg["center"](*args)
+            else:
+                self._center = cfg["center"]
+        else:
+            self._center = None
+        self._rvs_transform = cfg.get("rvs_transform", None)
+        self._rvs_transform_inv = cfg.get("rvs_transform_inv", None)
+        _mirror_uniform = cfg.get("mirror_uniform", None)
```
I did not understand this comment. "mirror_uniform" is just required for two distributions, so I did not want to add this key for all distributions

tupui replied:
It was a general comment for the whole section. I am questioning the architecture to create a `CustomDistPINV`.

chrisb83 replied:
Initially I was considering using a factory method design pattern but the current design is much simpler and avoids a lot of redundancy.

chrisb83 commented on scipy/stats/_sampling.py:
```diff
@@ -0,0 +1,1213 @@
+
+        Notes
+        -----
+        Note that the default ``rvs`` method of the rv_continuous class is
+        overwritten. Hence, a different stream of random numbers is generated
+        even if the same seed is used.
```
do you have  a strong view about removing this comment? I would say it does not hurt to mention this, we have a similar comment in the sampling module. otherwise I can also drop it

we cannot use `NumericalInversePolynomial.rvs` directly if mirror_uniform is true.

tupui replied:
It might not hurt (not sure), but I also don't think it's helpful. It confused me at first as I had to think about this comment. Ideally end users should not really care about the existence of `rv_continuous` and how it relates to it. This is what I meant, I see this more as a dev comment and it could go besides the code.

chrisb83 replied:
I updated the entire docstring to give more context, it was not clear before since I copied it from the generic rvs method in the sampling submodule

chrisb83 commented:
@tupui Thanks for the review, I replied to most of your comments. I will try to update this PR soon

chrisb83 commented on scipy/stats/tests/test_fast_gen_inversion.py:
```diff
@@ -0,0 +1,450 @@
+        (4, (4,)),
+        ((4,), (4,)),
+        ((2, 4), (2, 4)),
+    ]  # type: ignore
```
the comment needs to be removed, it causes one of the remaining build failures

chrisb83 commented on scipy/stats/_sampling.py:
```diff
@@ -0,0 +1,1213 @@
+        Name of the distribution in `scipy.stats`. The list of supported
+        distributions can be found in the Notes section.
+    args : tuple of floats or empty tuple, optional
+        The shape parameters (without `loc` and `scale`) of the distribution.
```
I raise an error if the length of args is greater than the number of shape parameters.

chrisb83 commented on scipy/stats/_sampling.py:
```diff
@@ -0,0 +1,1213 @@
+        The default is None. In that case, the random variates are not
+        truncated, and the domain is inferred from the support of the
+        distribution.
+    loc : float, optional
+        The location parameter. Note that only floats and no arrays are
+        allowed. The default is 0.
+    scale : float, optional
+        The scale parameter. Note that an array of floatsis not allowed.
+        The default is 1.
```
I thought it is cleaner and more explicit with a keyword, plus it is in line with the `rv_continuous` methods that separate the shape parameters and loc/scale

not having loc and scale would be very problematic since it creates an inconsistency to how parameters are defined in SciPy, e.g., you need `scale` to define the exponential distribution with shape `scale`

chrisb83 commented on scipy/stats/_sampling.py:
```diff
@@ -0,0 +1,1213 @@
+                if not cfg["check_pinv_params"](*args):
+                    msg = "No generator is defined for the shape parameters "
+                    msg += f"{args}. Use ignore_shape_range to proceed "
+                    msg += "with the selected values."
```
done

chrisb83 commented:
@tupui I have tried to addressed all your points whenever I replied to them above with "agree", "done" or similar :) the points that require further clarification or a decision about the API are not changed. feel free to resolve some of the conversations if you are happy with the updates. thx!

tupui commented:
Thanks @chrisb83 for the updates. I resolved some conversations and left open the rest for more comments from others.

I am still -1 on a few points, main ones being `args` (name and tuple like) and to expose all these methods (I am still convinced it should just have `[q]rvs` and `evaluate_error`. As it is, to my liking it's too close to another distribution interface.)

I don't think my opinion will change on these, so I would prefer to have other maintainer's giving their opinions and taking up the responsibility to go with this.

chrisb83 commented:
> Thanks @chrisb83 for the updates. I resolved some conversations and left open the rest for more comments from others.
> I am still -1 on a few points, main ones being `args` (name and tuple like) and to expose all these methods (I am still convinced it should just have `[q]rvs` and `evaluate_error`. As it is, to my liking it's too close to another distribution interface.)

Thanks, I can rename `args` to something like `shape_args` or `shape_params`. I would not use a dict though since naming the parameters seems too complicated (usually just one or no shape parameter, two parameter in a few cases). 

@tirthasheshpatel @mdhaber do you have a view on the API (see comment https://github.com/scipy/scipy/pull/17955#discussion_r1102973217,  https://github.com/scipy/scipy/pull/17955#pullrequestreview-1390000295)?

chrisb83 commented on scipy/stats/sampling.py:
```diff
@@ -49,3 +70,1262 @@
+    ----------
+    distname : str
+        Name of the distribution in `scipy.stats`.
+    args : tuple of floats or empty tuple, optional
```
I had a chat with Matt about this topic. A simple way would be to change the interface as follows:
it accepts a frozen distribution object as an input: it contains the name, the parameters incl loc/scale. So the API will become simpler w/o changing much under the hood (for example, the dictionary with the settings can be used as before). You probably discussed it with Matt already. If you agree, I can change it accordingly

chrisb83 commented:
@tupui I have changed the API as outlined in option 2 in https://github.com/scipy/scipy/pull/17955#discussion_r1186440397

please let me know what you think

chrisb83 commented:
The Linux Meson 3.11 failures are unrelated. 
The job "Linux 32bit" has crashed twice after/during (?) `test_sampling.py`. This is something to investigate

tylerjereddy commented:
From discussing with Pamphile and Matt in person, it sounds like this is likely technically correct, but perhaps a bit of nervousness about exposing it in the public API.

It sounds like, at a very minimum, this needs to hit the mailing list again because it has changed substantially, before it can be merged. At the same time, the feeling is that there is some really good work here, so although I'm bumping the milestone, if folks feel in favor it would be great if they speak up to make it easier for us to decide to move forward with this soon.

chrisb83 commented:
> The job "Linux 32bit" has crashed twice after/during (?) test_sampling.py. This is something to investigate

I tried to reproduce the error by running the tests on the docker image that is used by this workflow. All stats tests passed w/o errors. When I commit more changes, I will merge main to see if this goes away.

chrisb83 commented:
@tupui please let me know your thoughts on how to continue with this PR. I think the changes should have addressed the points you raised. thanks!

tupui reviewed (commented):
Thank you for the update Christoph. The API is more contained now which is a plus to me 👍 

I don't remember you discussed the why this is a class and not a function "augmenting" a frozen distribution (i.e. overwriting the methods defined here). I know that this was part of the last talk we had with Matt.

I am still not sure about `ppf_fast` and would just name it `ppf`. Ultimately if a user is using that, they are interested in the fast version.

Conflicts would need to be resolved (and that should fix the CI issues on CircleCI), but otherwise I think it's in good shape and we could move forward.

chrisb83 commented:
> I don't remember you discussed the why this is a class and not a function "augmenting" a frozen distribution (i.e. overwriting the methods defined here). I know that this was part of the last talk we had with Matt.

One key feature is to support sampling from truncated distributions which cannot be done easily otherwise. And since that is not supported by the current infrastructure, I cannot rely on pdf, cdf etc of the frozen distribution. 
 
> I am still not sure about `ppf_fast` and would just name it `ppf`. Ultimately if a user is using that, they are interested in the fast version.

True, we also call it `ppf` in NumericalInverseP... and NumericalInverseH... and since we dropped the "original" ppf, let's rename it as you suggested. I will change that when I fix the conflicts.

tupui commented:
To be clear, the alternative I am talking about would not return a normal frozen distribution but it would completely overwrite some of its methods/attributes.

But the negative point of that is that it would be change the semantic of the object and might not be clear then for a user what object they are dealing with.

tupui reviewed (approved):
Modulo the last changes (`ppf` and finish the discussion around frozen dists), this is good to go and we can follow up (mostly docs with figs as discussed.)

tupui commented:
I did the `ppf` change, if the CI comes back green I will merge 😃

tupui commented:
@chrisb83 there is actually an issue with the 32 bits tests which is related to this PR. We don't have the logs on Azure anymore but the 32 bits job seem to have been failing for a while on this PR.


```
scipy/stats/tests/test_sampling.py ..................................... [ 96%]
..
Error: Process completed with exit code 1.
```

The error is not helpful at all so I suppose there is an issue with UNURAN behind the scene.

tirthasheshpatel commented:
> The error is not helpful at all so I suppose there is an issue with UNURAN behind the scene.

Yep, UNU.RAN's 32-bit support is not great. I have encountered this before and decided to just `xfail` the test. UNU.RAN segfaults sometimes for the 32-bit tests since some methods require support for 64-bit integers. I will try to debug what's happening here though, I didn't expect to see this again.

chrisb83 commented:
Yeah, this jobs fails consistently. I tried to reproduce it by pulling the docker image that the 32bit CI job relies on, but the tests all passed... https://github.com/scipy/scipy/pull/17955#issuecomment-1561814503

I can also check it again. @tirthasheshpatel thanks for offering to help!

chrisb83 commented:
> I did the `ppf` change

Thanks!

chrisb83 commented:
I tried to reproduce it on Linux 32 bit and ran the stats tests several times, it never failed. I had downloaded the image quay.io/pypa/manylinux2014_i686 a while ago and then followed the steps in https://github.com/scipy/scipy/blob/main/.github/workflows/linux_meson.yml

```
docker run  -i -t --platform=linux/i386 f92da6a7efd8
cd /home
git clone https://github.com/chrisb83/scipy.git
cd scipy
git checkout unuran_sampling
git submodule update --init
python3.9 tools/openblas_support.py
cd ..
cp -r /tmp/openblas/lib/* /usr/local/lib
cp /tmp/openblas/include/* /usr/local/include
export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig
python3.9 -m venv test
source test/bin/activate
python -m pip install doit click rich_click pydevtool meson ninja
python -m pip install numpy==1.22.4 cython==0.29.35 pybind11 pytest pytest-timeout pytest-xdist pytest-env 'Pillow<10.0.0' mpmath pythran pooch meson


cd scipy
LD_LIBRARY_PATH=/usr/local/lib python dev.py build
LD_LIBRARY_PATH=/usr/local/lib python dev.py test -v -s stats
```

tupui commented:
I've noticed that in `test_sampling.py` we had a skip with this reason: `NumericalInversePolynomial.qrvs fails for Win 32-bit`. So trying to add this now on the tests doing `qrvs` in `test_fast_gen_inversion.py`
